File: solutions/go/proverb/1/proverb.go
package proverb

import "fmt"

func Proverb(rhyme []string) (r []string) {
    size := len(rhyme)
    if size == 1 {
        r = append(r, fmt.Sprintf("And all for the want of a %s.", rhyme[0]))
    }
	for i, v := range rhyme {
        if i == 0 {
            continue
        }
        r = append(r, fmt.Sprintf("For want of a %s the %s was lost.", rhyme[i-1], v))
        if i == size-1 {
            r = append(r, fmt.Sprintf("And all for the want of a %s.", rhyme[0]))
        }
    }
    return
}

File: solutions/python/proverb/1/proverb.py
from itertools import chain

def proverb(*args, qualifier = None):
    return [
        line for line in chain((
            f'For want of a {first} the {second} was lost.' for first, second in zip(
                args[:-1], args[1:])
        ), ('And all for the want of a ' + ' '.join(
            filter(lambda e: e, (qualifier, args[0]))
        ) + '.',))
    ] if args else []